File: api_sample/web/media-http-stream/nx-media-client.mjs
// Copyright 2018-present Network Optix, Inc. Licensed under MPL 2.0: www.mozilla.org/MPL/2.0/
/**
 * Play an Nx camera's live (or archive) video in an HTML5 <video> tag — from
 * the BROWSER, in BOTH auth modes: "Direct to Media Server" (connect to one
 * media server by IP:port with a LOCAL server account) or "Pull Stream via
 * Cloud Relay" (a cloud account reaching the site over the relay).
 *
 * Browser/front-end sample on the latest /rest/v4 API. As with every other web
 * sample, the pure, framework-free logic lives here so it can be (a) imported
 * by the page (app.mjs) and (b) tested offline with node:test and a fake fetch.
 * The DOM wiring is all in app.mjs.
 *
 * THE MEDIA ENDPOINT (from the v4 spec):
 *
 *   GET /rest/v4/devices/{id}/media.webm   (Authorization: Bearer <token>)
 *     ?stream=primary|secondary   hi/lo-res; we default to secondary (lighter)
 *     &positionMs=<ms>            archive start time; OMIT this for live
 *     &durationMs=<ms>           optional clip length
 *
 *   webm is the container an HTML5 <video> can play progressively for live.
 *
 * THE KEY DESIGN — authenticating a <video> tag (read the README):
 *
 *   A `<video src>` is a plain GET the browser issues; it CANNOT carry an
 *   Authorization header. So the flow is:
 *
 *     1. The page logs in THROUGH THE PROXY (a normal fetch, which CAN send/read
 *        JSON) and gets a bearer token in JS — login() below.
 *     2. The page sets <video>.src to a SAME-ORIGIN proxy URL that carries the
 *        token as a query param `auth=<token>` — buildMediaUrl() below.
 *     3. The PROXY strips `auth` off the forwarded URL and re-sends it to Nx as
 *        `Authorization: Bearer <token>`, then STREAMS the video body back.
 *
 *   So the token only ever travels SAME-ORIGIN to the local proxy; it is never
 *   sent to Nx as a URL parameter. (See proxy.mjs.)
 *
 * WHY THE BROWSER ALWAYS GOES THROUGH THE PROXY (read the README):
 *
 *   The cloud, the relay, and a local server are all different origins from this
 *   page and do NOT send CORS headers for it, so the browser blocks direct
 *   calls. A local server also presents a self-signed TLS cert the browser
 *   refuses. The included proxy.mjs serves this page AND relays calls
 *   same-origin, so the client just uses relative routes:
 *
 *        {baseUrl}/cloud/...                       -> the cloud      (cloud login)
 *        {baseUrl}/relay/<siteId>/...            -> the site relay (cloud media)
 *        {baseUrl}/server/<encoded-base>/...    -> the user's media server
 *                                                   (direct login + media)
 *
 *   baseUrl defaults to "" (same origin = the proxy that served the page). For
 *   the DIRECT mode the first /server/ path segment is the URI-encoded media
 *   server base URL (scheme+host+port) the user typed on the page, e.g.
 *   /server/https%3A%2F%2F192.168.1.10%3A7001/rest/v4/... — the proxy decodes it
 *   and forwards there. For cloud mode the proxy also performs the relay's
 *   307 + bearer re-attach hop, which a browser cannot do.
 */

export const CLIENT_ID = "3rdParty";
// API version path segment. v4 is the latest Nx REST API.
export const API = "/rest/v4";

// The two auth modes this sample supports.
export const MODE_DIRECT = "direct";
export const MODE_CLOUD = "cloud";

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

export class AuthError extends Error {
  constructor(message) {
    super(message);
    this.name = "AuthError";
  }
}

export class ApiError extends Error {
  constructor(message) {
    super(message);
    this.name = "ApiError";
  }
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

export class NxMediaClient {
  /**
   * @param {object} cfg
   * @param {"direct"|"cloud"} cfg.mode   Auth mode.
   * @param {string} cfg.user       LOCAL server username (direct) / cloud email (cloud).
   * @param {string} cfg.password   LOCAL server password (direct) / cloud password (cloud).
   * @param {string} [cfg.serverAddress]  DIRECT mode only: the media server base
   *        URL the user typed on the page — scheme+host+port, e.g.
   *        "https://192.168.1.10:7001". The proxy forwards there.
   * @param {string} [cfg.siteId]    Cloud Site ID (UUID). Required for cloud mode.
   * @param {string|null} [cfg.mfaCode]   Cloud 2FA code (cloud mode only).
   * @param {string} [cfg.baseUrl]  Origin that serves the proxy routes.
   *        Defaults to "" = same origin (the proxy that served this page).
   *        You should not need to set this.
   * @param {function} [cfg.fetchImpl] Injected for offline tests. Defaults to
   *        the global fetch (browser or Node 18+).
   */
  constructor({ mode, user, password, serverAddress = "", siteId = "", mfaCode = null, baseUrl = "", fetchImpl = null }) {
    this.mode = mode;
    this.user = user;
    this.password = password;
    this.serverAddress = (serverAddress || "").replace(/\/+$/, "");
    this.siteId = siteId;
    this.mfaCode = mfaCode;
    this.baseUrl = (baseUrl || "").replace(/\/+$/, "");
    // Default to the global fetch, but call it through a wrapper so it keeps
    // its `window`/global receiver. Calling `this.fetchImpl(...)` where
    // fetchImpl IS window.fetch throws "Can only call Window.fetch on
    // instances of Window" in browsers — the bound wrapper avoids that.
    this.fetchImpl = fetchImpl || ((...args) => globalThis.fetch(...args));
    this.token = null;
  }

  /** Same-origin route the proxy forwards to the cloud. */
  get cloudUrl() {
    return `${this.baseUrl}/cloud`;
  }

  /** Same-origin route the proxy forwards to this site's relay. */
  get relayUrl() {
    return `${this.baseUrl}/relay/${this.siteId}`;
  }

  /**
   * Same-origin route the proxy forwards to the user-provided media server.
   * The first path segment is the URI-encoded server base URL the user typed
   * (scheme+host+port), e.g. /server/https%3A%2F%2F192.168.1.10%3A7001. The
   * proxy decodes it and forwards there.
   */
  get serverUrl() {
    if (!this.serverAddress) {
      throw new ApiError("A server address is required for Direct to Media Server mode.");
    }
    return `${this.baseUrl}/server/${encodeURIComponent(this.serverAddress)}`;
  }

  /** The same-origin base the media URL is built on (mode-dependent). */
  get mediaBase() {
    return this.mode === MODE_CLOUD ? this.relayUrl : this.serverUrl;
  }

  // -------------------------------------------------------------------------
  // login(): get a bearer token. Two flows, one method.
  // -------------------------------------------------------------------------

  async login() {
    return this.mode === MODE_CLOUD ? this._loginCloud() : this._loginDirect();
  }

  /** Direct: POST {server}/rest/v4/login/sessions -> { token }. */
  async _loginDirect() {
    const url = `${this.serverUrl}${API}/login/sessions`;
    const body = {
      username: this.user,
      password: this.password,
      // A token (not a cookie) is what we want for a Bearer-header flow.
      setCookie: false,
    };

    let response;
    try {
      response = await this.fetchImpl(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      });
    } catch (exc) {
      throw new ApiError(
        `Could not reach the API at ${url}: ${exc.message}. ` +
          "Is the dev server running (node server.mjs) and is the Server address " +
          "you entered correct and reachable? — see the README.",
      );
    }
    if (response.status === 401 || response.status === 403) {
      throw new AuthError(
        `Login rejected (HTTP ${response.status}): the server did not accept this ` +
          "username/password. Verify they're a LOCAL server account (cloud users " +
          "use the Pull Stream via Cloud Relay mode).",
      );
    }
    if (!response.ok) {
      const text = await safeText(response);
      throw new ApiError(`Login failed: HTTP ${response.status} ${text.slice(0, 200)}`);
    }
    let data;
    try {
      data = await response.json();
    } catch {
      throw new ApiError("Login response was not valid JSON.");
    }
    this.token = data.token;
    if (!this.token) throw new ApiError("Login response did not contain a token.");
    return this.token;
  }

  /** Cloud: POST {cloud}/cdb/oauth2/token with cloudSystemId scope -> { access_token }. */
  async _loginCloud() {
    const url = `${this.cloudUrl}/cdb/oauth2/token`;
    const body = {
      grant_type: "password",
      response_type: "token",
      client_id: CLIENT_ID,
      username: this.user,
      password: this.password,
      // THIS scope is what makes the token usable against the site relay.
      scope: `cloudSystemId=${this.siteId}`,
    };
    if (this.mfaCode) body.mfaCode = this.mfaCode;

    let response;
    try {
      response = await this.fetchImpl(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      });
    } catch (exc) {
      throw new ApiError(
        `Could not reach the API at ${url}: ${exc.message}. ` +
          "Is the dev server running? (node server.mjs) — see the README.",
      );
    }
    if (response.status === 401 || response.status === 403) {
      throw new AuthError(
        `Login rejected (HTTP ${response.status}): the cloud did not accept this ` +
          "email/password. Verify they're exactly what you use at the Nx Cloud " +
          "portal (nxvms.com). (2FA accounts also need an MFA code.)",
      );
    }
    if (!response.ok) {
      const text = await safeText(response);
      throw new ApiError(`Token request failed: HTTP ${response.status} ${text.slice(0, 200)}`);
    }
    let data;
    try {
      data = await response.json();
    } catch {
      throw new ApiError("Token response was not valid JSON.");
    }
    this.token = data.access_token;
    if (!this.token) throw new ApiError("Token response did not contain an access_token.");
    return this.token;
  }

  // -------------------------------------------------------------------------
  // buildMediaUrl(): the same-origin <video> src the proxy will authenticate.
  // -------------------------------------------------------------------------

  /**
   * Build the SAME-ORIGIN media URL to put on <video>.src. The token rides as
   * an `auth` query param ONLY because a <video> GET can't send a header; the
   * proxy strips it and re-sends it as `Authorization: Bearer <token>` to Nx,
   * so the token never reaches Nx as a URL param. See proxy.mjs.
   *
   * @param {object} opts
   * @param {string} opts.deviceId          Camera/device id (the {id} path part).
   * @param {"primary"|"secondary"} [opts.stream]  Default "secondary".
   * @param {number|null} [opts.positionMs] Archive start in epoch ms; OMIT for live.
   * @param {number|null} [opts.durationMs] Optional clip length in ms.
   * @returns {string} e.g.
   *   /server/<encoded-base>/rest/v4/devices/<id>/media.webm?stream=...&auth=...
   */
  buildMediaUrl({ deviceId, stream = "secondary", positionMs = null, durationMs = null }) {
    if (!this.token) throw new ApiError("Not logged in. Call login() first.");
    if (!deviceId) throw new ApiError("A deviceId is required to build the media URL.");

    const path = `${this.mediaBase}${API}/devices/${encodeURIComponent(deviceId)}/media.webm`;
    const params = new URLSearchParams();
    params.set("stream", stream);
    // positionMs present == archive; absent == live. Treat null/"" as live.
    if (positionMs !== null && positionMs !== undefined && positionMs !== "") {
      params.set("positionMs", String(positionMs));
    }
    if (durationMs !== null && durationMs !== undefined && durationMs !== "") {
      params.set("durationMs", String(durationMs));
    }
    // The proxy will REMOVE this and turn it into an Authorization header.
    params.set("auth", this.token);
    return `${path}?${params.toString()}`;
  }

  // -------------------------------------------------------------------------
  // logout(): revoke the token. Best-effort cleanup.
  // -------------------------------------------------------------------------

  async logout() {
    if (!this.token) return;
    const url =
      this.mode === MODE_CLOUD
        ? `${this.cloudUrl}/cdb/oauth2/token/${this.token}`
        : `${this.serverUrl}${API}/login/sessions/${this.token}`;
    try {
      await this.fetchImpl(url, {
        method: "DELETE",
        headers: { Authorization: `Bearer ${this.token}` },
      });
    } catch {
      // best effort
    } finally {
      this.token = null;
    }
  }
}

async function safeText(response) {
  try {
    return await response.text();
  } catch {
    return "";
  }
}

// ---------------------------------------------------------------------------
// Parsing the archive position the user types (ISO time or epoch ms).
// ---------------------------------------------------------------------------

/**
 * Turn the optional "archive position" field into epoch milliseconds, or null
 * for live. Accepts an ISO 8601 string (2026-06-15T12:00:00Z) or a raw epoch-ms
 * number. Empty/blank -> null (live).
 *
 * @returns {number|null}
 */
export function parsePositionMs(value) {
  const s = value === undefined || value === null ? "" : String(value).trim();
  if (!s) return null; // live
  // All-digits -> already epoch ms.
  if (/^\d+$/.test(s)) return Number(s);
  const ms = Date.parse(s);
  if (Number.isNaN(ms)) {
    throw new ApiError(`Could not parse archive position "${s}". Use ISO time or epoch ms.`);
  }
  return ms;
}

// ---------------------------------------------------------------------------
// Config: the fields the user types. In Direct mode the user also types the
// media server address (scheme+host+port) on the page.
// ---------------------------------------------------------------------------

export function resolveConfig(values = {}) {
  const v = (key) => (values[key] === undefined || values[key] === null ? "" : String(values[key]).trim());
  const mode = v("mode") === MODE_CLOUD ? MODE_CLOUD : MODE_DIRECT;
  return {
    mode,
    serverAddress: v("serverAddress"),
    user: v("user"),
    password: v("password"),
    siteId: v("siteId"),
    mfaCode: v("mfaCode") || null,
    deviceId: v("deviceId"),
    stream: v("stream") === "primary" ? "primary" : "secondary",
    position: v("position"), // raw; parsed with parsePositionMs() at play time
  };
}

/**
 * Which required fields are missing for the chosen mode. Direct needs
 * serverAddress+user+password+deviceId; cloud needs siteId+user+password+deviceId.
 */
export function missingFields(config) {
  const required =
    config.mode === MODE_CLOUD
      ? ["siteId", "user", "password", "deviceId"]
      : ["serverAddress", "user", "password", "deviceId"];
  return required.filter((k) => !config[k]);
}

File: api_sample/csharp/media-http-stream/src/NxMediaClient.cs
// Copyright 2018-present Network Optix, Inc. Licensed under MPL 2.0: www.mozilla.org/MPL/2.0/
// Save a short video CLIP from an Nx camera to a FILE (C#) — the command-line
// counterpart of the browser ../../web/media-http-stream sample. A CLI can't
// render a <video>, so instead it fetches the media stream and writes it to a
// file you can play in VLC / ffplay / a browser.
//
// C# port of ../../typescript/media-http-stream, on the latest /rest/v4 API.
// Uses the built-in HttpClient + System.Text.Json — no third-party packages.
//
// BOTH auth modes, exactly like the browser/TypeScript sample:
//
//   --mode direct  Direct to Media Server: connect to ONE media server by
//                  IP:port with a LOCAL server account.
//                    NX_SERVER_HOST / NX_SERVER_USER / NX_SERVER_PASSWORD
//                  Login:  POST {server}/rest/v4/login/sessions
//                          { username, password, setCookie:false }  ->  { token }
//   --mode cloud   Pull Stream via Cloud Relay: a cloud account reaches the
//                  site over the relay (token scoped with cloudSystemId, the
//                  relay 307 followed manually with the bearer re-attached).
//                    NX_CLOUD_HOST / NX_CLOUD_USER / NX_CLOUD_PASSWORD / NX_CLOUD_SITE_ID
//                  Login:  POST {cloud}/cdb/oauth2/token
//                          { grant_type:"password", response_type:"token",
//                            client_id:"3rdParty", username, password,
//                            scope:"cloudSystemId=<site id>" }  ->  { access_token }
//
// THE MEDIA ENDPOINT (from the v4 spec):
//
//   GET /rest/v4/devices/{id}/media.{format}   (Authorization: Bearer <token>)
//     ?positionMs=<ms>     archive start time; OMIT this for LIVE
//     &durationMs=<ms>     how much footage to pull (bounds the clip)
//
//   format is one of the containers the v4 spec allows for this endpoint (see
//   Formats, taken verbatim from docs/v4_api_spec.json):
//     webm, mpegts, mpjpeg, mp4, mkv, _3gp, rtp, flv, f4v
//
// LIVE vs ARCHIVE:
//   - No --pos            -> LIVE: save the next --duration seconds.
//   - --pos <ISO|epochMs> -> ARCHIVE: save --duration seconds starting there.
//
// Because a CLI must terminate, the clip is always bounded by --duration
// (seconds, default 10). durationMs is sent to the server AND, via a
// CancellationToken with a timeout of durationMs + grace, used as a client-side
// safety stop so the program can never hang on an endless live stream.
//
// Streaming: the response is read with HttpCompletionOption.ResponseHeadersRead
// and copied straight from response.Content to a destination Stream, so the
// whole clip is never buffered in memory. The destination is injectable (any
// Stream) so tests don't need disk; a real FileStream is used by the CLI.
//
// Relay 307: the relay answers with a 307 redirect to the node that serves the
// request. .NET's HttpClient (like browsers) drops the Authorization header on a
// cross-host redirect, so we disable auto-redirect and follow it ourselves,
// re-attaching the bearer on each hop. See GetFollowingRedirectsAsync.

using System.Globalization;
using System.Net;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace NxMediaHttpStream;

public sealed class AuthException : Exception
{
    public AuthException(string message) : base(message) { }
}

public sealed class ApiException : Exception
{
    public ApiException(string message) : base(message) { }
}

/// <summary>The two auth modes this sample supports (same names as the web sample).</summary>
public enum Mode
{
    Direct,
    Cloud,
}

/// <summary>What a single saveClip call needs to know about the media request.</summary>
public sealed record ClipRequest(string DeviceId, string Format, long? PositionMs, long? DurationMs);

public sealed class NxMediaClient
{
    public const string ClientId = "3rdParty";
    public const string RelaySuffix = ".relay.vmsproxy.com";
    public const string Api = "/rest/v4";

    // Container formats the v4 media.{format} endpoint supports, copied verbatim
    // from the `format` enum in docs/v4_api_spec.json. Don't invent others.
    public static readonly IReadOnlyList<string> Formats = new[]
    {
        "webm", "mpegts", "mpjpeg", "mp4", "mkv", "_3gp", "rtp", "flv", "f4v",
    };
    public const string DefaultFormat = "webm";

    // Clip length when --duration is not given (seconds).
    public const int DefaultDurationSeconds = 10;
    // Extra wall-clock grace beyond durationMs before the client-side abort fires.
    private const int AbortGraceMs = 10000;
    // Most redirects we will follow when chasing the relay 307.
    private const int MaxRedirects = 5;

    private readonly HttpClient _http;
    private readonly Mode _mode;
    private readonly string _user;
    private readonly string _password;
    private readonly string _serverHost;
    private readonly string _cloudHost;
    private readonly string _siteId;
    private readonly string? _mfaCode;

    public string? Token { get; private set; }

    public NxMediaClient(
        HttpClient http,
        Mode mode,
        string user,
        string password,
        string? serverHost = null,
        string? cloudHost = null,
        string? siteId = null,
        string? mfaCode = null)
    {
        _http = http;
        _mode = mode;
        _user = user;
        _password = password;
        _serverHost = (serverHost ?? "").TrimEnd('/');
        _cloudHost = (cloudHost ?? "https://nxvms.com").TrimEnd('/');
        _siteId = siteId ?? "";
        _mfaCode = mfaCode;
    }

    /// <summary>The Cloud relay address for this specific site (cloud mode).</summary>
    public string RelayUrl => $"https://{_siteId}{RelaySuffix}";

    /// <summary>Where media requests go: the server directly, or the site relay.</summary>
    public string MediaBase => _mode == Mode.Cloud ? RelayUrl : _serverHost;

    /// <summary>Use a bearer token obtained elsewhere (skip login).</summary>
    public void UseToken(string token) => Token = token;

    // -----------------------------------------------------------------------
    // LoginAsync(): two flows, one method.
    // -----------------------------------------------------------------------

    public Task<string> LoginAsync(CancellationToken cancellationToken = default)
        => _mode == Mode.Cloud ? LoginCloudAsync(cancellationToken) : LoginDirectAsync(cancellationToken);

    /// <summary>Direct: POST {server}/rest/v4/login/sessions -> { token }.</summary>
    private async Task<string> LoginDirectAsync(CancellationToken cancellationToken)
    {
        string url = $"{_serverHost}{Api}/login/sessions";
        var body = new Dictionary<string, object>
        {
            ["username"] = _user,
            ["password"] = _password,
            ["setCookie"] = false,
        };

        using var content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");

        HttpResponseMessage response;
        try
        {
            response = await _http.PostAsync(url, content, cancellationToken);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
        {
            throw new ApiException($"Could not reach {url}: {ex.Message}");
        }

        using (response)
        {
            if (response.StatusCode is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
            {
                throw new AuthException(
                    $"Login rejected (HTTP {(int)response.StatusCode}). Check the username/password, "
                    + "and that it is a LOCAL server account (cloud users use --mode cloud).");
            }
            if (!response.IsSuccessStatusCode)
            {
                string text = await SafeReadAsync(response);
                throw new ApiException(
                    $"Login failed: HTTP {(int)response.StatusCode} {Truncate(text, 200)}");
            }
            string json = await response.Content.ReadAsStringAsync(cancellationToken);
            Token = ExtractToken(json);
            return Token;
        }
    }

    /// <summary>Cloud: POST {cloud}/cdb/oauth2/token with cloudSystemId scope -> { access_token }.</summary>
    private async Task<string> LoginCloudAsync(CancellationToken cancellationToken)
    {
        string url = $"{_cloudHost}/cdb/oauth2/token";
        var body = new Dictionary<string, string>
        {
            ["grant_type"] = "password",
            ["response_type"] = "token",
            ["client_id"] = ClientId,
            ["username"] = _user,
            ["password"] = _password,
            // THIS scope is what makes the token usable against the site relay.
            ["scope"] = $"cloudSystemId={_siteId}",
        };
        if (!string.IsNullOrEmpty(_mfaCode)) body["mfaCode"] = _mfaCode;

        using var content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");

        HttpResponseMessage response;
        try
        {
            response = await _http.PostAsync(url, content, cancellationToken);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
        {
            throw new ApiException($"Could not reach {url}: {ex.Message}");
        }

        using (response)
        {
            if (response.StatusCode is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
            {
                throw new AuthException(
                    $"Login rejected (HTTP {(int)response.StatusCode}). Check the cloud email/password, "
                    + "the site id, and that the account has access to that site. Add --mfa-code for 2FA.");
            }
            if (!response.IsSuccessStatusCode)
            {
                string text = await SafeReadAsync(response);
                throw new ApiException(
                    $"Token request failed: HTTP {(int)response.StatusCode} {Truncate(text, 200)}");
            }
            string json = await response.Content.ReadAsStringAsync(cancellationToken);
            Token = ExtractAccessToken(json);
            return Token;
        }
    }

    // -----------------------------------------------------------------------
    // BuildMediaUrl(): the upstream media URL (header auth — no token in the URL).
    // -----------------------------------------------------------------------

    public string BuildMediaUrl(string deviceId, string format, long? positionMs = null, long? durationMs = null)
    {
        if (string.IsNullOrEmpty(deviceId))
        {
            throw new ApiException("A deviceId is required to build the media URL.");
        }
        string path = $"{MediaBase}{Api}/devices/{Uri.EscapeDataString(deviceId)}/media.{format}";
        var parts = new List<string>();
        // positionMs present == archive; absent == live.
        if (positionMs is not null)
        {
            parts.Add($"positionMs={Uri.EscapeDataString(positionMs.Value.ToString(CultureInfo.InvariantCulture))}");
        }
        if (durationMs is not null)
        {
            parts.Add($"durationMs={Uri.EscapeDataString(durationMs.Value.ToString(CultureInfo.InvariantCulture))}");
        }
        return parts.Count == 0 ? path : $"{path}?{string.Join("&", parts)}";
    }

    // -----------------------------------------------------------------------
    // SaveClipAsync(): fetch the media stream and copy it to `destination`.
    // -----------------------------------------------------------------------

    /// <summary>
    /// Fetch the clip and STREAM the response body into <paramref name="destination"/>,
    /// returning the number of bytes written. Uses
    /// HttpCompletionOption.ResponseHeadersRead and copies the content stream
    /// directly, so the whole clip is never buffered in memory. A
    /// CancellationToken with a timeout of durationMs + grace stops an endless
    /// live stream so the CLI can never hang.
    /// </summary>
    public async Task<long> SaveClipAsync(
        Stream destination, ClipRequest request, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrEmpty(Token))
        {
            throw new ApiException("Not logged in. Call LoginAsync() or UseToken() first.");
        }

        string url = BuildMediaUrl(request.DeviceId, request.Format, request.PositionMs, request.DurationMs);

        using var timeoutCts = new CancellationTokenSource();
        using var linked = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCts.Token);
        if (request.DurationMs is > 0)
        {
            // Client-side safety stop: never wait forever on an endless stream.
            timeoutCts.CancelAfter(TimeSpan.FromMilliseconds(request.DurationMs.Value + AbortGraceMs));
        }

        using HttpResponseMessage response = await GetFollowingRedirectsAsync(url, linked.Token);

        if (response.StatusCode is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
        {
            throw new AuthException(
                $"The server rejected the token (HTTP {(int)response.StatusCode}). In cloud mode make "
                + "sure it was scoped with cloudSystemId for THIS site.");
        }
        if (!response.IsSuccessStatusCode)
        {
            string text = await SafeReadAsync(response);
            throw new ApiException(
                $"Media request failed: HTTP {(int)response.StatusCode} {Truncate(text, 200)}");
        }

        return await CopyToCountingAsync(response, destination, linked.Token);
    }

    /// <summary>
    /// Copy the response content stream to <paramref name="destination"/> without
    /// buffering the whole clip, returning the byte count.
    /// </summary>
    private static async Task<long> CopyToCountingAsync(
        HttpResponseMessage response, Stream destination, CancellationToken cancellationToken)
    {
        Stream source;
        try
        {
            source = await response.Content.ReadAsStreamAsync(cancellationToken);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException or IOException)
        {
            throw new ApiException($"Media response had no body to save: {ex.Message}");
        }

        long total = 0;
        var buffer = new byte[81920];
        int read;
        while ((read = await source.ReadAsync(buffer.AsMemory(0, buffer.Length), cancellationToken)) > 0)
        {
            await destination.WriteAsync(buffer.AsMemory(0, read), cancellationToken);
            total += read;
        }
        await destination.FlushAsync(cancellationToken);
        return total;
    }

    // -----------------------------------------------------------------------
    // GetFollowingRedirectsAsync(): the relay 307, followed by hand.
    // -----------------------------------------------------------------------

    /// <summary>
    /// GET that follows 307 redirects MANUALLY, re-attaching the bearer on each
    /// hop so it survives the relay's cross-host redirect, and reads only the
    /// response headers (ResponseHeadersRead) so the body is streamed, not
    /// buffered. (Requires the HttpClient's handler to have AllowAutoRedirect = false.)
    /// </summary>
    private async Task<HttpResponseMessage> GetFollowingRedirectsAsync(
        string url, CancellationToken cancellationToken)
    {
        string current = url;
        for (int hop = 0; hop <= MaxRedirects; hop++)
        {
            using var request = new HttpRequestMessage(HttpMethod.Get, current);
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", Token);

            HttpResponseMessage response;
            try
            {
                response = await _http.SendAsync(
                    request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
            }
            catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
            {
                throw new ApiException($"Could not reach {current}: {ex.Message}");
            }

            int code = (int)response.StatusCode;
            if (code is 301 or 302 or 303 or 307 or 308)
            {
                Uri? location = response.Headers.Location;
                response.Dispose();
                if (location is null)
                {
                    throw new ApiException($"Redirect {code} without a Location header.");
                }
                current = location.IsAbsoluteUri ? location.ToString() : new Uri(new Uri(current), location).ToString();
                continue; // re-send WITH the bearer header
            }
            return response;
        }
        throw new ApiException($"Too many redirects (>{MaxRedirects}) chasing the relay.");
    }

    // -----------------------------------------------------------------------
    // LogoutAsync(): revoke the token. Best-effort cleanup.
    // -----------------------------------------------------------------------

    public async Task LogoutAsync(CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrEmpty(Token)) return;
        string url = _mode == Mode.Cloud
            ? $"{_cloudHost}/cdb/oauth2/token/{Token}"
            : $"{_serverHost}{Api}/login/sessions/{Token}";
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Delete, url);
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", Token);
            using HttpResponseMessage _ = await _http.SendAsync(request, cancellationToken);
        }
        catch
        {
            // best effort
        }
        finally
        {
            Token = null;
        }
    }

    // -----------------------------------------------------------------------
    // Parsing helpers (pure = easy to test)
    // -----------------------------------------------------------------------

    public static string ExtractToken(string json)
    {
        JsonDocument doc;
        try { doc = JsonDocument.Parse(json); }
        catch (JsonException) { throw new ApiException("Login response was not valid JSON."); }
        using (doc)
        {
            if (doc.RootElement.ValueKind == JsonValueKind.Object
                && doc.RootElement.TryGetProperty("token", out JsonElement el)
                && el.ValueKind == JsonValueKind.String)
            {
                return el.GetString()!;
            }
            throw new ApiException("Login response did not contain a token.");
        }
    }

    public static string ExtractAccessToken(string json)
    {
        JsonDocument doc;
        try { doc = JsonDocument.Parse(json); }
        catch (JsonException) { throw new ApiException("Token response was not valid JSON."); }
        using (doc)
        {
            if (doc.RootElement.ValueKind == JsonValueKind.Object
                && doc.RootElement.TryGetProperty("access_token", out JsonElement el)
                && el.ValueKind == JsonValueKind.String)
            {
                return el.GetString()!;
            }
            throw new ApiException("Token response did not contain an access_token.");
        }
    }

    private static async Task<string> SafeReadAsync(HttpResponseMessage response)
    {
        try { return await response.Content.ReadAsStringAsync(); }
        catch { return string.Empty; }
    }

    private static string Truncate(string s, int max) => s.Length <= max ? s : s[..max];
}
